### PR TITLE
Implement RFC 45: Move `hdl.Memory` to `lib.Memory`.

### LINKED ISSUE #216: Memory port transparency model is flawed
<a href="https://github.com/whitequark"></img></a> **Issue by [whitequark](https://github.com/whitequark)**
_Friday Sep 20, 2019 at 18:04 GMT_
_Originally opened as https://github.com/m-labs/nmigen/issues/216_

----

Investigating #172, it appears that there are two major issues with the way port transparency is handled.
  1. Transparent read ports should not be the default, since it is expensive. On many platforms (e.g. iCE40), fabric logic will need to be inserted. On other platforms (e.g. Xilinx), non-transparent ports have a much larger set of configurations with well-defined behavior. Transparency between different clock domains is impossible with BRAMs and will result in a (silent) failure during synthesis.
  2. Transparency should not be requested *in general*, but rather for a specific write port (in the same domain). It does not make sense to request a "fully" transparent read port for a true dual-port RAM with write ports in different domains. It does make sense to request a read port transparent only wrt one specific write port for a true dual-port RAM with write ports in the same domain (this is implementable in Xilinx) and a read port transparent wrt both write ports for a true dual-port RAM with write ports in the same domain (this is not directly implementable in any FPGA arch I know, but is useful for e.g. superscalar CPU register files).

I propose to remove `transparent` completely, replacing it with a `transparent_for=[write_ports]` argument. The default (when this argument is not specified) would be "transparent for no write ports" for synchronous read ports, and "transparent for all ports" for asynchronous read ports. It would not be possible to specify a non-default value for an asynchronous read port.

The lowering to RTLIL would be the same as today, with the selected write port ignored (beyond checking that the domain is compatible),

### PR DESCRIPTION
**Pull request by wanda-phi**
Fixes #216.

### CODE REVIEW

File: amaranth/lib/memory.py [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+        
+        def __repr__(self):
+            return f"Memory.Init({self._items!r})"
+
+    def __init__(self, *, depth, shape, init, attrs=None, src_loc_at=0, src_loc=None):
+        # shape and depth validation performed in Memory.Init constructor.
+        self._depth = depth
+        self._shape = shape
+        self._init = Memory.Init(init, shape=shape, depth=depth)
+        self._attrs = OrderedDict(() if attrs is None else attrs)
```
Comment (whitequark) [inline amaranth/lib/memory.py:365]: `OrderedDict` is just a normal dict since Python 3.6.

File: amaranth/lib/memory.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+        self.src_loc = src_loc or tracer.get_src_loc(src_loc_at=1 + src_loc_at)
```
Comment (whitequark) [inline amaranth/lib/memory.py:366]: Is this the right `src_loc_at`?

File: amaranth/lib/memory.py [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+                if shape.length == 0:
+                    en_width = 0
+                elif granularity == 0:
+                    raise ValueError("Granularity must be positive")
+                elif shape.length % granularity != 0:
+                    raise ValueError("Granularity must divide data array length")
+                else:
+                    en_width = shape.length // granularity
+            else:
+                raise TypeError("Granularity can only be specified for plain `Shape` or `ArrayLayout`")
```
Comment (whitequark) [inline amaranth/lib/memory.py:94]: `unsigned`, not just plain `Shape`, right?

**Review by whitequark — COMMENTED** [review]
Did we decide to get rid of `MemoryIdentity` in the end?

**Comment by wanda-phi** [comment]
> Did we decide to get rid of `MemoryIdentity` in the end?

I ended up including it, since this allows us to recognize memories of all kinds (old and new) as "something that has an `_identity` that is `MemoryInstance`" in the simulator, avoiding a dependency.

Still not sure if this is the way to go.

File: amaranth/lib/memory.py (hunk 2 above)
Comment (wanda-phi) [reply amaranth/lib/memory.py:366]: whoops, good catch

**Review by whitequark — COMMENTED** [review]
Can you back out the docs rst file? I'd like to improve these before publishing, but I don't want to block the PR on that.

**Comment by wanda-phi** [comment]
> Can you back out the docs rst file? I'd like to improve these before publishing, but I don't want to block the PR on that.

Done

**Review by whitequark — APPROVED** [review]
Thanks!

**Pull request merged**